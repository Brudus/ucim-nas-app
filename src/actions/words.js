import database from '../firebase/firebase';

// ADD_WORD
export const addWord = (word) => ({
    type: 'ADD_WORD',
    word
});

export const startAddWord = (wordData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            source = '', 
            destination= '', 
            repeatAt = 0,
            interval = 0,
            reps = 0,
            easeFactor = 2.5,
            repeatAtInverted = 0,
            intervalInverted = 0,
            repsInverted = 0,
            easeFactorInverted = 2.5
        } = wordData;
        const word = { source, destination, repeatAt, interval, reps, easeFactor, repeatAtInverted, intervalInverted, repsInverted, easeFactorInverted };
        
        return database.ref(`users/${uid}/words`).push(word).then((ref) => {
            dispatch(addWord({
                id: ref.key,
                ...word
            }));
        });
    };
};

// REMOVE_WORD
export const removeWord = ({ id } = {}) => ({
    type: 'REMOVE_WORD',
    id
});

export const startRemoveWord = ({ id } = {}) => {
    return (dispatch,getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/words/${id}`).remove().then(() => {
            dispatch(removeWord({ id }));
        });
    };
};

// EDIT_WORD
export const editWord = (id, updates) => ({
    type: 'EDIT_WORD',
    id,
    updates
});

export const startEditWord = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/words/${id}`).update(updates).then(() => {
            dispatch(editWord(id, updates));
        })
    };
};

// SET_WORDS
export const setWords = (words) => ({
    type: 'SET_WORDS',
    words
});

export const startSetWords = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/words`).once('value').then((snapshot) => {
            const words = [];

            snapshot.forEach((childSnapshot) => {
                words.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            
            dispatch(setWords(words));
        });
    };
};