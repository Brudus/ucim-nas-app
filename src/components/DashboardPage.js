import React from 'react';

const DashboardPage = () => (
    <div>
        <p>You are logged in!</p>
    </div>
);

export default DashboardPage;