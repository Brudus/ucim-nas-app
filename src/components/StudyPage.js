import React from 'react';
import { connect } from 'react-redux';

export const StudyPage = (props) => (
    <div>
        <h2>Study!</h2>
        {props.words[0].source}
    </div>
);

const mapStateToProps = (state) => ({
    words: state.words
});

export default connect(mapStateToProps)(StudyPage);