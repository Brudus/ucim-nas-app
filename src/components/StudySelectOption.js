import React from 'react';

export default class StudySelectionOption extends React.Component {
    handleClick = () => {
        this.props.calculateNewSchedule(this.props.grade);
    };

    render() {
        return (
            <div>
                <button onClick={this.handleClick}>{this.props.text}</button>
            </div>
        );
    };
};