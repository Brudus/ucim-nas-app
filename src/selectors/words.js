export default (words, { text, sortBy }) => {
    return words.filter((word) => {
        const sourceText = word.source.toLowerCase()
        const destinationText = word.destination.toLowerCase();
        const searchText = text.toLowerCase().trim();

        const sourceMatch = sourceText.includes(searchText);
        const destinationMatch = destinationText.includes(searchText);
        return sourceMatch || destinationMatch;
    }).sort((a, b) => {
        if (sortBy === 'repeatAt') {
            return a.repeatAt > b.repeatAt ? 1 : -1;
        } else if (sortBy === 'source') {
            const aText = a.source.toLowerCase().trim();
            const bText = b.source.toLowerCase().trim();
            return aText > bText ? 1 : -1;
        } else if (sortBy === 'destination') {
            const aText = a.destination.toLowerCase().trim();
            const bText = b.destination.toLowerCase().trim();
            return aText > bText ? 1 : -1;
        }
    });
};