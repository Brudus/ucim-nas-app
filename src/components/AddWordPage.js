import React from 'react';
import { connect } from 'react-redux';
import WordForm from './WordForm';
import { startAddWord } from '../actions/words';

export class AddWordPage extends React.Component {
    onSubmit = (word) => {
        this.props.startAddWord(word);
    };

    render() {
        return (
            <div>
                <h1>Add Word</h1>
                <WordForm 
                    onSubmit={this.onSubmit}
                />
            </div>
        );
    };
}

const mapDispatchToProps = (dispatch) => ({
    startAddWord: (word) => dispatch(startAddWord(word))
});

export default connect(undefined, mapDispatchToProps)(AddWordPage);