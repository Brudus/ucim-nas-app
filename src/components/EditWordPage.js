import React from 'react';
import { connect } from 'react-redux';
import WordForm from './WordForm';
import { startEditWord, startRemoveWord } from '../actions/words';

export class EditWordPage extends React.Component {
    onSubmit = (word) => {
        this.props.startEditWord(this.props.word.id, word);
        this.props.history.push('/catalog');
    };
    onRemove = (id) => {
        this.props.startRemoveWord({ id: this.props.word.id });
        this.props.history.push('/catalog');
    };

    render() {
        return (
            <div>
                <WordForm 
                    word={this.props.word}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onRemove}>Remove</button>
            </div>
        );
    };
};

const mapStateToProps = (state, props) => ({
    word: state.words.find((word) => word.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
    startEditWord: (id, word) => dispatch(startEditWord(id, word)),
    startRemoveWord: (data) => dispatch(startRemoveWord(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditWordPage);