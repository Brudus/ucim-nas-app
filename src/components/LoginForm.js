import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export class LoginForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            email: '',
            password: '',
            loginError: '',
            emailPassed: undefined
        };
    }

    validateEmail = (email) => {
        const regex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
        return regex.test(email);
    };

    onEmailChange = (e) => {
        const email = e.target.value;
        const emailPassed = this.validateEmail(email);
        this.setState(() => ({ 
            email,
            emailPassed
        }));
    };

    onPasswordChange = (e) => {
        const password = e.target.value;
        this.setState(() => ({ password }));
    }

    onSubmit = (e) => {
        e.preventDefault();

        this.setState(() => ({ 
            loginError: '',
            emailPassed: true
        }));
        this.props.startLogin(this.state.email, this.state.password);
    };

    render() {
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    {this.state.loginError && <p>{this.state.loginError}</p>}
                    <div>
                        <input 
                            type="email"
                            placeholder="Email address" 
                            autoFocus
                            onChange={this.onEmailChange}
                        />
                    </div>
                    
                    <div>
                        <input 
                            type="password"
                            placeholder="Password" 
                            onChange={this.onPasswordChange}
                        />
                    </div>
                    <button disabled={!this.state.emailPassed || !this.state.password}>Login</button>
                </form> 
            </div>
        );
    };
};

const mapDispatchToProps = (dispatch, props) => ({
    startLogin: (email, password) => dispatch(startLogin(email, password))
});

export default connect(undefined, mapDispatchToProps)(LoginForm);