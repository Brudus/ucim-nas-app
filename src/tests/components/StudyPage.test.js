import React from 'react';
import { shallow } from 'enzyme';
import words from '../fixtures/words';
import { StudyPage } from '../../components/StudyPage';

const startDate = 1487076708000;
const dayInMs = 1000 * 60 * 60 * 24;
const globalDate = Date;
let startEditWord, history, wordsWOReference;

beforeAll(() => {
    Date.now = jest.fn(() => startDate);
});

beforeEach(() => {
    startEditWord = jest.fn();
    history = { push: jest.fn() };
    wordsWOReference = JSON.parse(JSON.stringify(words));
});

afterAll(() => {
    global.Date = globalDate;
});

test('should render StudyPage with words', () => {
    const wrapper = shallow(
        <StudyPage 
            startEditWord={startEditWord} 
            history={history} 
            words={wordsWOReference} 
        />
    );
    expect(wrapper.state('showAnswer')).toBe(false);
    expect(wrapper.state('currentWord')).toEqual(words[0]);
    expect(wrapper.state('words')).toEqual(words);
    expect(wrapper).toMatchSnapshot();
});

test('should push on history if words are empty', () => {
    const wrapper = shallow(
        <StudyPage 
            startEditWord={startEditWord} 
            history={history} 
            words={[]} 
        />
    );
    expect(history.push).toHaveBeenLastCalledWith('/'); 
});

test('should set answer state on button click', () => {
    const wrapper = shallow(
        <StudyPage 
            startEditWord={startEditWord} 
            history={history} 
            words={wordsWOReference} 
        />
    );
    wrapper.find('button').simulate('click');
    expect(wrapper.state('showAnswer')).toBe(true);
    expect(wrapper).toMatchSnapshot();
});

test('should calculate schedule correctly and put word at end of list for grade 0', () => {
    const wrapper = shallow(
        <StudyPage 
            startEditWord={startEditWord} 
            history={history} 
            words={wordsWOReference} 
        />
    );
    wrapper.find('button').simulate('click');
    const stateCurrentWord = wrapper.state('currentWord');
    wrapper.find('StudySelectionOption').at(0).prop('calculateNewSchedule')(0);
    expect(stateCurrentWord.interval).toBe(0);
    expect(stateCurrentWord.reps).toBe(0);
    expect(stateCurrentWord.easeFactor).toBe(words[0].easeFactor - 0.8);
    expect(stateCurrentWord.repeatAt).toEqual(1487076708000);
    expect(wrapper.state('currentWord')).toEqual(words[1]);
    expect(wrapper.state('showAnswer')).toBe(false);
    expect(wrapper.state('words')).toEqual([words[1], words[2], stateCurrentWord]);
    expect(startEditWord).toHaveBeenLastCalledWith(stateCurrentWord.id, {
        repeatAt: stateCurrentWord.repeatAt,
        interval: stateCurrentWord.interval,
        reps: stateCurrentWord.reps,
        easeFactor: stateCurrentWord.easeFactor
    }); 
});

test('should calculate schedule correctly and put word at end of list for grade 3', () => {
    const wrapper = shallow( 
        <StudyPage 
            startEditWord={startEditWord} 
            history={history} 
            words={wordsWOReference} 
        />
    );
    wrapper.find('button').simulate('click');
    const stateCurrentWord = wrapper.state('currentWord');
    wrapper.find('StudySelectionOption').at(0).prop('calculateNewSchedule')(3);
    expect(stateCurrentWord.interval).toBe(0);
    expect(stateCurrentWord.reps).toBe(words[0].reps + 1);
    expect(stateCurrentWord.easeFactor).toBe(2.36);
    expect(stateCurrentWord.repeatAt).toEqual(1487076708000);
    expect(wrapper.state('currentWord')).toEqual(words[1]);
    expect(wrapper.state('showAnswer')).toBe(false);
    expect(wrapper.state('words')).toEqual([words[1], words[2], stateCurrentWord]);
    expect(startEditWord).toHaveBeenLastCalledWith(stateCurrentWord.id, {
        repeatAt: stateCurrentWord.repeatAt,
        interval: stateCurrentWord.interval,
        reps: stateCurrentWord.reps,
        easeFactor: stateCurrentWord.easeFactor
    });
});

test('should calculate schedule correctly and remove word from list for grade 4', () => {
    const wrapper = shallow( 
        <StudyPage 
            startEditWord={startEditWord} 
            history={history} 
            words={wordsWOReference} 
        />
    );
    wrapper.find('button').simulate('click');
    const stateCurrentWord = wrapper.state('currentWord');
    wrapper.find('StudySelectionOption').at(0).prop('calculateNewSchedule')(4);
    expect(stateCurrentWord.interval).toBe(8);
    expect(stateCurrentWord.reps).toBe(words[0].reps + 1);
    expect(stateCurrentWord.easeFactor).toBe(words[0].easeFactor);
    expect(stateCurrentWord.repeatAt).toEqual(1487076708000 + stateCurrentWord.interval * dayInMs);
    expect(wrapper.state('currentWord')).toEqual(words[1]);
    expect(wrapper.state('showAnswer')).toBe(false);
    expect(wrapper.state('words')).toEqual([words[1], words[2]]);
    expect(startEditWord).toHaveBeenLastCalledWith(stateCurrentWord.id, {
        repeatAt: stateCurrentWord.repeatAt,
        interval: stateCurrentWord.interval,
        reps: stateCurrentWord.reps,
        easeFactor: stateCurrentWord.easeFactor
    });
});

test('should calculate schedule correctly and remove word from list for grade 5', () => {
    const wrapper = shallow( 
        <StudyPage 
            startEditWord={startEditWord} 
            history={history} 
            words={wordsWOReference} 
        />
    );
    wrapper.find('button').simulate('click');
    const stateCurrentWord = wrapper.state('currentWord');
    wrapper.find('StudySelectionOption').at(0).prop('calculateNewSchedule')(5);
    expect(stateCurrentWord.interval).toBe(8);
    expect(stateCurrentWord.reps).toBe(words[0].reps + 1);
    expect(stateCurrentWord.easeFactor).toBe(2.6);
    expect(stateCurrentWord.repeatAt).toEqual(1487076708000 + stateCurrentWord.interval * dayInMs);
    expect(wrapper.state('currentWord')).toEqual(words[1]);
    expect(wrapper.state('showAnswer')).toBe(false);
    expect(wrapper.state('words')).toEqual([words[1], words[2]]);
    expect(startEditWord).toHaveBeenLastCalledWith(stateCurrentWord.id, {
        repeatAt: stateCurrentWord.repeatAt,
        interval: stateCurrentWord.interval,
        reps: stateCurrentWord.reps,
        easeFactor: stateCurrentWord.easeFactor
    });
});

test('should calculate schedule correctly and put word at end of list for incorrect grade -1', () => {
    const wrapper = shallow(
        <StudyPage 
            startEditWord={startEditWord} 
            history={history} 
            words={wordsWOReference} 
        />
    );
    wrapper.find('button').simulate('click');
    const stateCurrentWord = wrapper.state('currentWord');
    wrapper.find('StudySelectionOption').at(0).prop('calculateNewSchedule')(7);
    expect(stateCurrentWord.interval).toBe(words[0].interval);
    expect(stateCurrentWord.reps).toBe(words[0].reps);
    expect(stateCurrentWord.easeFactor).toBe(words[0].easeFactor);
    expect(stateCurrentWord.repeatAt).toEqual(words[0].repeatAt);
    expect(wrapper.state('currentWord')).toEqual(words[0]);
    expect(wrapper.state('words')).toEqual(words); 
});

test('should calculate schedule correctly and put word at end of list for incorrect grade 7', () => {
    const wrapper = shallow(
        <StudyPage 
            startEditWord={startEditWord} 
            history={history} 
            words={wordsWOReference} 
        />
    );
    wrapper.find('button').simulate('click');
    const stateCurrentWord = wrapper.state('currentWord');
    wrapper.find('StudySelectionOption').at(0).prop('calculateNewSchedule')(7);
    expect(stateCurrentWord.interval).toBe(words[0].interval);
    expect(stateCurrentWord.reps).toBe(words[0].reps);
    expect(stateCurrentWord.easeFactor).toBe(words[0].easeFactor);
    expect(stateCurrentWord.repeatAt).toEqual(words[0].repeatAt);
    expect(wrapper.state('currentWord')).toEqual(words[0]);
    expect(wrapper.state('words')).toEqual(words); 
});

test('should navigate to / when last word is learned', () => {
    const wrapper = shallow(
        <StudyPage 
            startEditWord={startEditWord} 
            history={history} 
            words={[wordsWOReference[1]]} 
        />
    );
    wrapper.find('button').simulate('click');
    const stateCurrentWord = wrapper.state('currentWord');
    wrapper.find('StudySelectionOption').at(0).prop('calculateNewSchedule')(5);
    expect(history.push).toHaveBeenLastCalledWith('/'); 
});