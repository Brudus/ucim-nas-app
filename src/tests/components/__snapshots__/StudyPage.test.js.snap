// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render StudyPage with words 1`] = `
<div>
  <h2>
    Study!
  </h2>
  <p>
    gut
  </p>
  <button
    onClick={[Function]}
  >
    Show answer
  </button>
</div>
`;

exports[`should set answer state on button click 1`] = `
<div>
  <h2>
    Study!
  </h2>
  <p>
    gut
  </p>
  <p>
    dobro
  </p>
  <div>
    <StudySelectionOption
      calculateNewSchedule={[Function]}
      grade={0}
      key="0"
      text="Again"
    />
    <StudySelectionOption
      calculateNewSchedule={[Function]}
      grade={3}
      key="3"
      text="Hard"
    />
    <StudySelectionOption
      calculateNewSchedule={[Function]}
      grade={4}
      key="4"
      text="Good"
    />
    <StudySelectionOption
      calculateNewSchedule={[Function]}
      grade={5}
      key="5"
      text="Easy"
    />
  </div>
</div>
`;
